Feature/add resource page

pumpkiinbell opened the pull request:
## :bookmark_tabs: 제목

강의자료 페이지 개발

<img width="1551" alt="스크린샷 2021-05-02 오후 1 07 47" src="https://user-images.githubusercontent.com/19240202/116801736-688c8780-ab47-11eb-83da-e66fec846ff8.png">

## :heavy_check_mark: 셀프 체크리스트

> ~최소 1명 이상의 assign을 받아야만 Merge 가능합니다.~

- [x] Warning Message가 발생하지 않았나요?
- [x] Coding Convention을 준수했나요?
- [ ] 'npm run lint'나 'yarn lint'를 실행하였나요?

## :speech_balloon: 작업 내용

> 구현 내용 및 작업 했던 내역

- [x] 강의자료 페이지 퍼블리싱
- [x] 강의자료 페이지에서 사용되어지는 API 연결 (백엔드 API 수정 완료 후 수정할 것 있음. TODO 확인할 것)

## :construction: PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점

- 아직 개발이 끝나지 않았습니다. 재락이 형이 작업 내역 확인하기 편하도록 우선적으로 PR 날립니다.

- TODO
    - 백엔드 API 수정 시 유저가 이전에 클릭했던 강의자료 확인할 수 있도록 수정
    - 백엔드 API 수정 시 유저가 좋아요 클릭 했을 때 hit 수 감산 제대로 나오도록 수정
    - 강의자료 작성 모달 창 퍼블리싱 및 API 연결

pumpkiinbell committed:
fix: Modify FilterBox component

강의자료 페이지에 맞게 필터 박스 섹션을 수정했습니다.

pumpkiinbell committed:
feat: Connect department on resources page

전공 버튼 클릭 시 상태값 변경할 수 있도록 수정

pumpkiinbell committed:
fix: Modify SearchForm to handle resources page

검색 바에서 사용하는 함수를 강의자료 검색을 위해 수정

pumpkiinbell committed:
feat: Fetch resources on resource page

filter option에 맞게 강의자료 데이터 호출

pumpkiinbell committed:
fix: Initialize department state when user clicked same button

유저가 같은 버튼을 클릭할 시 기존에 들어있던 부서 상태 값을 초기화시킨다.

pumpkiinbell committed:
refactor: Erase useless code

pumpkiinbell committed:
fix: Add spinner when fetching

pumpkiinbell committed:
Merge branch 'develop' into feature/add_resource_page

pumpkiinbell committed:
feat: Create ResourceCard

~/resources 에서 resources 보여주기 위한 컴포넌트입니다.

pumpkiinbell committed:
refactor: Change function name using verb

pumpkiinbell committed:
chore: Add pushed icon

pumpkiinbell committed:
fix: Fix error property

pumpkiinbell committed:
feat: Connect api to use when user clicked thumb icon

pumpkiinbell committed:
fix: Set token as null when user not logged in

pumpkiinbell committed:
fix: Change limit of resources

pumpkiinbell committed:
fix: Fix error that lectures not loaded when user come back into lectures page

sonjaerock commented on src/api/resources.js:
```diff
@@ -0,0 +1,49 @@
+import axios from "axios";
+
+const setAxiosConfig = (accessToken = null, data = null) => {
```
axios config 설정은 디게 많은데 header에 token 설정하는걸로 함수명이 좀 그런듯
setTokenInHeader 뭐 이런식이 맞는듯

pumpkiinbell replied:
맞는말이네요 수정하겠습니다

sonjaerock commented on src/api/resources.js:
```diff
@@ -0,0 +1,49 @@
+    },
+  };
+
+  if (data !== null) config["data"] = data;
```
이런것도 data가 undefined로 온다던가 공백으로 올때의 처리도 있으면 좋을듯?
그리고 validation check는 try catch로 하는게 정석이긴함

pumpkiinbell replied:
오... 하나배워씀다 ㅎㅎ

sonjaerock commented on src/api/resources.js:
```diff
@@ -0,0 +1,49 @@
+    });
+    query = query.slice(0, -1);
+
+    const response = await axios.get(`/lecture-banks/search?${query}`);
```
보통 react에서는 axios보다 fetch많이 쓰지않나 아닌가..?

pumpkiinbell replied:
둘다 httprequest 생성하는걸로 알고있는데... fetch 써야만 하는 큰 이유가 있을까요??

sonjaerock commented on src/components/ResourceComponents/ResourceCard.js:
```diff
@@ -0,0 +1,221 @@
+const PUSHED_THUMB_URL =
+  "https://hangang-storage.s3.ap-northeast-2.amazonaws.com/assets/img/resourcepage/thumb_up_pushed.png";
+const NOT_PUSHED_THUMB_URL =
+  "https://hangang-storage.s3.ap-northeast-2.amazonaws.com/assets/img/resourcepage/thumb_up.png";
```
이런것도 앞에 도메인은 constants.js이런거 만들어서 상수로 지정해두고 쓰는게 좋음 
s3링크가 바뀌게되면 상수만 교체하면 되니깐 ㅇㅇ

pumpkiinbell replied:
음 생각해보니 컬러는 따로 파일을 만들었는데 저런 파일 uri들은 안만들었네요 수정할게요!!

sonjaerock commented on src/components/ResourceComponents/ResourceCard.js:
```diff
@@ -0,0 +1,221 @@
+ */
+const sliceString = (string, max) => {
+  if (string.length > max) {
+    string = string.slice(0, max) + "...";
```
말줄임을로 저렇게 js 로 처리하면 안됨
예를들어 해당 영역이 넓어졌을 때(window width가 늘어날 경우) 결국 저 max값에의해서 data 유실되서 공간이 남아있음에도 ...으로 표시됨, 
css로 처리하는 기술 있음 확인바람

pumpkiinbell replied:
아 그러게요!!! 수정해서 css로 핸들링하겠습니다

sonjaerock commented on src/components/ResourceComponents/ResourceCard.js:
```diff
@@ -0,0 +1,221 @@
+ * @param {string} accessToken A string of access_token of user token.
+ * @param {function} setHits A function to change state of hits.
+ */
+const pushHitIcon = async (
```
개인차긴한데 보통 개발단에서 push는 Array 에 원소 추가하거나, 푸시서버에 주로 씀
클릭에 대한 메소드 같은데 clickHitIcon 으로 나였으면 썼을듯

pumpkiinbell replied:
아하... 그러게요 수정할게요 이 리뷰는 진짜 좀 아차싶었네요 진하게 하나 배웠습니다 ㅋㅋ

sonjaerock commented on src/components/Shared/FilterBox.js:
```diff
@@ -155,15 +160,32 @@ const labelConverter = (key) => {
     if (type === "lecture") {
       dispatch(setLectureFilter({ key, value }));
+    } else {
+      dispatch(setResourcesFilter({ key, value }));
```
여기도 좀 아쉽긴하네 뭔가 setFilter로 통일시키고 Reducer에서 구분할 수 있으면 로직이 줄어들텐데 
lectureReducer, resourceReducer의 부모 클래스 만들어놓고 protectd setFliter 로 해두면 가능할 것 같긴한데 ㅋㅋ
아직 oop 적용을 많이 안해봤나봄?

sonjaerock replied:
생각해보니, dispatch라 상위 객체단에서 처리 했으면 가능하긴한데 굳이 그럴필요까지는 없긴할듯

pumpkiinbell replied:
ㅜㅜ oop 적용은 진짜 미숙합니다...

sonjaerock commented on src/store/modules/resources.js:
```diff
@@ -0,0 +1,107 @@
+      let { key, value } = action.payload;
+
+      switch (key) {
+        case "sort":
```
뭐 크게 상관은 없는데 reducer case 마다 주석 추가해두면 좋을듯 
각 case마다 뭐가 동작하는지 잘몰겠네

pumpkiinbell replied:
음 넵넵 확인했습니다

sonjaerock reviewed (commented):
쭉 봤는데 시간에 맞춰서 수용할건 수용하고 다음에 할건 다음에 처리하셈

pumpkiinbell reviewed (commented):
말씀해주신 것들 모두 수정하겠습니다~

pumpkiinbell committed:
fix: rename setAxiosConfig to setTokenInHeader

함수명이 부적합하다고 판단하여 수정하였습니다.

pumpkiinbell committed:
chore: move url constants on static folders

상수로 사용하는 url들을 static 공유 폴더에 옮겼습니다.

pumpkiinbell committed:
fix: rename pushHitIcon

함수 명으로 push가 적합하지 않다고 판단하여 수정하였습니다.

pumpkiinbell committed:
feat: Create resource write modal

pumpkiinbell committed:
chore: Add check logic that user rollback state if click x button

유저가 x 버튼을 누를 경우 작성하던 글 내용을 없애도록 함

pumpkiinbell committed:
chore: Create title section on form

pumpkiinbell committed:
chore: Create semester component

pumpkiinbell committed:
chore: Create Lecture search section

pumpkiinbell committed:
chore: Create category section

pumpkiinbell committed:
chore: Create content section

pumpkiinbell committed:
fix: Change value of overflow-y property

pumpkiinbell committed:
chore: Create material section

pumpkiinbell committed:
chore: Create logic to calculate material length and size

pumpkiinbell committed:
fix: Change decimal point

pumpkiinbell committed:
fix: rename resourcewrite container to resourcecreate container

write보다 create가 맞는 표현이라 생각하고 수정하였습니다.

pumpkiinbell committed:
chore: Connect write and cancel write form api

GET ~/lecture-bank/write, DELETE ~/lecture-bank/cancel/${id} 추가하였습니다.

pumpkiinbell committed:
chore: Connect file upload api

현재 다중파일 업로드에서 백엔드와 소통이 필요해  추후 수정이 필요합니다.
또 현재 특정 파일 업로드 시 에러가 발생하고 있습니다. 에러 케이스들 조사해서 수정이 필요합니다.

pumpkiinbell committed:
chore: Connect cancel uploaded file api

pumpkiinbell committed:
chore: Create lecture search box with only styling

pumpkiinbell committed:
chore: Create Lecture search box

pumpkiinbell committed:
chore: Connect submit api when user finished creating resource.

pumpkiinbell committed:
fix: Change form as default when user click close button

pumpkiinbell committed:
fix: Erase console.log

pumpkiinbell committed:
fix: Change id of files when multiple data come into

pumpkiinbell committed:
fix: Add isFetched dependencies on useEffect hook

사용자가 강의자료를 작성 완료 후 데이터를 볼 수 있도록 IsFetched를 의존성 목록에 추가하였습니다.
즉, 글 작성이 끝나면 isFetched를 false로 하고 해당 훅은 false 인 isFetched 를 감지하여 강의자료 api를 재요청합니다.

pumpkiinbell committed:
chore: Connect and set debounce at lecture search api

pumpkiinbell committed:
fix: Change default option of resource store

pumpkiinbell committed:
fix: Remove useless function call

백엔드 API 수정으로 deprecated 된 함수를 제거하였습니다

pumpkiinbell committed:
fix: Connect api when user pushed thumb icon

pumpkiinbell committed:
fix: Erase local state to manage on resource create container

글 작성 취소 버튼을 눌렀을 때 local 상태 값으로 인해 초기화가 제대로 되지 않는 에러가 있어서 수정하였습니다.
